### PR TITLE
docs: install typedoc for generating API docs

### LINKED ISSUE #6649: Theme protocol documentation
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Description

As we are starting to make another theme (and also allow users to create their own themes from scratch), we should make clear what a theme needs to implement at the bare minimum, and what props these components will receive.

This includes:

- Route components, mostly documented in https://docusaurus.io/docs/advanced/routing. These are a must because plugins call these directly.
- The tabs component, because the npm2yarn remark plugin depends on it.

And that's it! We just need to make sure that the props are stable. For now, this is ensured through typedefs, but in the future, if we no longer use module declarations (discussed in #6629), these implicit protocols will not be present. We need something more explicit than that.

Note that `themeConfig` is intentionally left out of this list. I strongly believe that, while `navbar` and `footer` and `prism` and stuff are useful, they are not _obligatory_. A theme should be allowed without them, or have a different API shape.

### Self-service

- [X] I'd be willing to address this documentation request myself.

### PR DESCRIPTION
**Pull request by Josh-Cena**
## Motivation

Close #6649. Since we have JSDoc, we can use TypeDoc to autogenerate these APIs.

I would investigate how we can make it emit nicer Markdown—probably need to eventually fork the plugin and make lots of customizations.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

https://deploy-preview-7060--docusaurus-2.netlify.app/docs/api-next

### CODE REVIEW

**Comment by slorber** [comment]
That looks interesting.

The problem is that most theme-common is currently not a public API surface. 

If we document this officially on our website then users will start using internal APIs and we'll be in trouble.

File: packages/docusaurus/src/server/plugins/configs.ts [hunk 1]
```diff
@@ -120,7 +120,7 @@ export async function loadPluginConfigs(
     // Site config should be the highest priority.
     ...standalonePlugins,
     ...standaloneThemes,
-  ];
+  ].filter(<T>(x: T | false | null | undefined): x is T => Boolean(x));
```
Comment (slorber) [inline packages/docusaurus/src/server/plugins/configs.ts:123]: That's a change I think we could do and merge in another PR asap

I think we have an issue for it btw or it was discussed in the past

**Comment by Josh-Cena** [comment]
Ugh, looks too hard to maintain in the long run. Let's just hand-sync the docs instead.
